certora/specs/AllowancesInvariant.spec

```
@@ -0,0 +1,95 @@
+    // Aave dispatch
+    function _.repay(address, uint256, uint256, address) external => HAVOC_ECF;
+    // Compound dispatch
+    function _.repayBorrowBehalf(address, uint256) external => HAVOC_ECF;
+    function _.underlying() external => mockUnderlying() expect address;
+    function _.supplyTo(address, address, uint256) external => HAVOC_ECF;
+    function _.baseToken() external => mockBaseToken() expect address;
+    function _.compoundV2RedeemEth(uint256,address) external => HAVOC_ECF;
+    // Paraswap dispatch
+    function _.set(bytes memory, uint256 offset, uint256) internal => setData(offset) expect void;
+    function _.get(bytes memory, uint256 offset) internal => getData(offset) expect uint256;
+    function _.isValidAugustus(address) external => NONDET;
+    // ERC20
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    unresolved external in _.nativeTransfer(address, uint256) => DISPATCH [] default HAVOC_ECF;
+    unresolved external in _.unwrapNative(uint256,address) => DISPATCH [] default HAVOC_ECF;
+    unresolved external in _.compoundV2RedeemEth(uint256, address) => DISPATCH [] default HAVOC_ECF;
+    unresolved external in _.onMorphoRepay(uint256, bytes) => DISPATCH [] default HAVOC_ECF;
+    unresolved external in _.onMorphoSupply(uint256, bytes) => DISPATCH [] default HAVOC_ECF;
+    unresolved external in _.onMorphoSupplyCollateral(uint256, bytes) => DISPATCH [] default HAVOC_ECF;
+    unresolved external in _.onMorphoFlashLoan(uint256, bytes) => DISPATCH [] default HAVOC_ECF;
+    unresolved external in _._ => DISPATCH [] default ASSERT_FALSE;
+}
+
+persistent ghost address lastErc20Underlying;
+persistent ghost bool erc20UnderlyingChanged;
+
+function mockUnderlying() returns address {
+    address erc20;
+    if (erc20 != lastErc20Underlying) {
+        erc20UnderlyingChanged = true;
+        lastErc20Underlying = erc20;
+    }
+    return erc20;
+}
+
+persistent ghost address lastErc20BaseToken;
+persistent ghost bool erc20BaseTokenChanged;
+
+function mockBaseToken() returns address {
+    address erc20;
+    if (erc20 != lastErc20BaseToken) {
+        erc20BaseTokenChanged = true;
+        lastErc20BaseToken = erc20;
+    }
+    return erc20;
+}
+
+persistent ghost mapping(uint256 => uint256) data;
+
+function getData(uint256 offset) returns uint256 {
+    return data[offset];
+}
+
+function setData(uint256 offset) {
+    havoc data;
+}
+
+// Ghost variable to store changed allowances.
+// This models only direct changes in allowances.
+persistent ghost mapping (address => mapping (address => uint256)) changedAllowances {
+    init_state axiom forall address token. forall address spender. changedAllowances[token][spender] == 0 ;
+}
+
+definition isKnownImmutable (address spender) returns bool =
```

```suggestion
definition isTrusted (address spender) returns bool =
```
because there are immutable contracts for which you don't want to give allowance